### PR TITLE
Add validator for IP overlap

### LINKED ISSUE #2077: No check for overlapping IP ranges
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a problem observed when **deploying** a Data Safe Haven.
- [x] I can reproduce this with the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :computer: System information

- **Operating System:** macOS
- **Data Safe Haven version:** develop @ 78ba7b7

## :package: Packages

<details>
<summary>List of packages</summary>

```none
Paste list of packages here
```

</details>

## :no_entry_sign: Describe the problem

We allow deployers to specify which IP addresses users/admin etc. will connect from but don't check whether these overlap. If they do, the NSG rules will fail to deploy.

Example of overlap:
- 1.1.0.0/24
- 1.1.0.22/32

## :deciduous_tree: Log messages

<details>
<summary>Relevant log messages</summary>

```none
  azure-native:network:NetworkSecurityGroup (sre_networking_nsg_application_gateway):
    error: Code="OverlappingSubnetsNotPermittedInSecurityRule" Message="Security rule parameter SourceAddressPrefix for rule with Id
/subscriptions/3f1a8e26-eae2-4539-952a-0a6184ec248a/resourceGroups/shm-pink-sre-fuschia-rg/providers/Microsoft.Network/networkSecurityGroups/shm-pink-sre-fuschia-nsg-application-gateway/securityRules/AllowUsersInternetInbound contains
overlapping subnets [91.235.65.0/24, 91.235.65.22/32], which is not permitted." Details=[]
```

</details>

## :recycle: To reproduce

Try to deploy an SRE with overlapping ranges in `admin_ip_addresses`

### PR DESCRIPTION
**Pull request by JimMadge**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :closed_umbrella: Related issues

Closes #2077

### CODE REVIEW

File: data_safe_haven/config/config_sections.py [hunk 1]
```diff
@@ -59,3 +60,11 @@ class ConfigSectionSRE(BaseModel, validate_assignment=True):
     software_packages: SoftwarePackageCategory = SoftwarePackageCategory.NONE
     timezone: TimeZone = "Etc/UTC"
     workspace_skus: list[AzureVmSku] = Field(..., default_factory=list[AzureVmSku])
+
+    @fieldvalidator("admin_ip_addresses", "data_provider_ip_addresses", "research_user_ip_addresses", mode="after")
+    @classmethod
+    def ensure_non_overlapping(cls, v: list[IpAddress]) -> list[IpAddress]:
```
Comment (jemrobinson) [inline data_safe_haven/config/config_sections.py:73]: This doesn't account for overlaps like `1.1.0.0/24` and `1.1.0.20/32`. This probably needs the functions in `IPv4Network` from `ipaddress`.

**Review by jemrobinson — CHANGES_REQUESTED** [review]

File: data_safe_haven/config/config_sections.py (hunk 1 above)
Comment (JimMadge) [reply data_safe_haven/config/config_sections.py:73]: Ah, thought that there were ip addresses not strings.

Comment (jemrobinson) [reply data_safe_haven/config/config_sections.py:73]: They are strings, but they're CIDRs. You will need to convert them to `IPv4Network`s then check for overlap.

**Review by jemrobinson — APPROVED** [review]
LGTM

**Pull request merged**